web/source/kmwembedded.ts

```diff
@@ -361,14 +362,25 @@ namespace com.keyman.text {
       /* Clear any pending (non-popup) key */
       osk.vkbd.keyPending = null;
 
-      // Changes for Build 353 to resolve KMEI popup key issues      
+      // Changes for Build 353 to resolve KMEI popup key issues
       keyName=keyName.replace('popup-',''); //remove popup prefix if present (unlikely)
 
-      // Can't just split on '-' because some layers like ctrl-shift contain it.
-      let separatorIndex = keyName.lastIndexOf('-');
+      let displayLayer = keymanweb.core.keyboardProcessor.layerId;
 
-      if (separatorIndex > 0) {
-        keyName = keyName.substring(separatorIndex+1);
+      // Regex for 'display layer'-'virtual key name'+'optional functional layer'
+      // Can't just split on '-' because some layers like ctrl-shift contain it.
+      // Virtual key can be T_, U_, K_, or ISO 9995
+      let rx = new RegExp("(.*)-([TKU]_[^+]*|[A-E]\\d\\d)\\+?(.*)?");
```

Comment:
Six things:

1. This regex should explicitly match the entire identifier -- we need to include `^` and `$`.
2. Why not use the normal Javascript regex enclosing with `/`, which avoids double-escaping?
3. Rather than use `.*` for the layer components, you should use `.+` -- we don't want a zero-length layer identifier!
4. I've removed the ISO9995 component.
5. A key identifier must be `T_`, `K_` or `U_` with one-or-more, not zero-or-more chars.
6. Use `const` where possible (or eliminate variables).

```suggestion
      const rx = /^(.+)-([TKU]_[^+]+)\+?(.+)?$/;
```

Or:

```
      const matches = keyName.match(/^(.+)-([TKU]_[^+]+)\+?(.+)?$/);
```